### PR TITLE
STM32 HSEM Id list update and rework impacted drivers

### PR DESCRIPTION
**Pull request by ABOSTM**
STM32 HSEM Id list update and rework impacted drivers.

Due to STM32WB Cube update, HSEM Id list has been updated. It introduces new semaphores for flash operations.
This PR contains 3 commits:
* soc: arm: st_stm32: update HSEM ID from STM32WB cube update
* drivers: STM32: Rework CLK48 HSEM protection 
* drivers: flash: stm32wb: rework dualcore flash operation

### CODE REVIEW

File: drivers/flash/flash_stm32wbx.c [hunk 1]
```diff
@@ -62,15 +67,105 @@ static int write_dword(struct device *dev, off_t offset, uint64_t val)
...
+		 * If the PESD is set, the CPU1 will be stalled when reading
+		 * literals from an ISR that may occur after the flash
+		 * processing has been requested but suspended due to the PESD
+		 * bit.
+		 *
+		 * Note: This code is required only when the PESD mechanism is
+		 * used to protect the CPU2 timing.
+		 * However, keeping that code make it compatible with both
+		 * mechanisms.
+		 */
```
Comment (gmarull) [inline drivers/flash/flash_stm32wbx.c:88]: This looks like copy from https://github.com/STMicroelectronics/STM32CubeWB/blob/master/Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_RfWithFlash/Core/Src/flash_driver.c#L185, code which is not Apache licensed. Not sure if that could be a problem? @erwango

Comment (erwango) [reply drivers/flash/flash_stm32wbx.c:88]: This is indeed an issue. We're triggering discussions to sort this out.

Comment (ABOSTM) [reply drivers/flash/flash_stm32wbx.c:89]: After internal FOSS Out process, agreement is reached to allow this code to be published with Apache licence.
I simply added STMicroelectronics copyright.

**Review by erwango — APPROVED** [review]

**Comment by ABOSTM** [comment]
@gmarull @ydamigos @superna9999 @loicpoulain 
Can you please review?

File: drivers/flash/flash_stm32wbx.c [hunk 2]
```diff
@@ -97,17 +197,100 @@ static int erase_page(struct device *dev, uint32_t page)
...
```
Comment (gmarull) [inline drivers/flash/flash_stm32wbx.c:218]: Not a blocker, but I think that mechanism details (i.e. all detailed comments) could go in a single place as a sort of documentation. In my opinion there's too much comments inline with the code.

File: drivers/flash/flash_stm32wbx.c [hunk 3]
```diff
@@ -97,17 +197,100 @@ static int erase_page(struct device *dev, uint32_t page)
...
+				 * sector.
+				 * When several sectors need to be erased,
+				 * the application shall first exit from the
+				 * critical section and try again.
+				 */
+				regs->CR |= FLASH_CR_PER;
+				regs->CR &= ~FLASH_CR_PNB_Msk;
+				regs->CR |= page << FLASH_CR_PNB_Pos;
+
+				regs->CR |= FLASH_CR_STRT;
```
Comment (gmarull) [inline drivers/flash/flash_stm32wbx.c:270]: Does LL API offer any API call for doing that?

File: drivers/flash/flash_stm32wbx.c (hunk 2 above)
Comment (ABOSTM) [reply drivers/flash/flash_stm32wbx.c:218]: As code and comments are inherited from https://github.com/STMicroelectronics/STM32CubeWB/blob/master/Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_RfWithFlash/Core/Src/flash_driver.c
I would prefer to keep it as is. Especially for further maintenance.

File: drivers/flash/flash_stm32wbx.c (hunk 3 above)
Comment (ABOSTM) [reply drivers/flash/flash_stm32wbx.c:270]: No LL API for that.

**Review by gmarull — APPROVED** [review]

**Comment by ABOSTM** [comment]
@carlescufi , @nashif , @galak,
Could this PR be merged (for next release) ?

**Review by loicpoulain — APPROVED** [review]
LGTM

**Review by carlescufi — APPROVED** [review]

**Pull request merged**